homeassistant/components/homeassistant_hardware/firmware_config_flow.py

```diff
@@ -313,6 +322,58 @@ async def _configure_and_start_otbr_addon(self) -> None:
 
         await otbr_manager.async_start_addon_waiting()
 
+        # Push the preferred Thread dataset to the OTBR addon after it starts
+        await self._push_dataset_to_otbr()
+
+    async def _push_dataset_to_otbr(self) -> None:
+        """Push the preferred Thread dataset to the OTBR addon after migration."""
+        otbr_manager = get_otbr_addon_manager(self.hass)
+        addon_hostname = hostname_from_addon_slug(otbr_manager.addon_slug)
+
+        # Wait for hassio discovery to create or update an OTBR config entry
+        otbr_entry_future: asyncio.Future[OTBRConfigEntry] = (
+            asyncio.get_running_loop().create_future()
+        )
+
+        @callback
+        def handle_config_entry_change(
+            change_type: ConfigEntryChange, entry: ConfigEntry
+        ) -> None:
+            """Handle config entry changes to find our OTBR entry."""
+            if (
+                entry.domain == OTBR_DOMAIN
+                and entry.state is ConfigEntryState.LOADED
+                and f"http://{addon_hostname}:" in entry.data["url"]
+                and not otbr_entry_future.done()
+            ):
+                otbr_entry_future.set_result(cast(OTBRConfigEntry, entry))
+
+        unsub = async_dispatcher_connect(
+            self.hass, SIGNAL_CONFIG_ENTRY_CHANGED, handle_config_entry_change
+        )
+
+        try:
+            async with asyncio.timeout(OTBR_INTEGRATION_SETUP_TIMEOUT):
+                otbr_entry = await otbr_entry_future
+        except TimeoutError:
+            _LOGGER.warning("Timed out waiting for the OTBR integration to start")
+            return
+        finally:
+            unsub()
+
+        thread_dataset_tlv = await async_get_preferred_dataset(self.hass)
+
+        # If we have no preferred dataset, OTBR will have started up with random
+        # settings so we can just use those
+        if thread_dataset_tlv is None:
+            return
+
+        _LOGGER.debug("Pushing preferred Thread dataset to OTBR")
+        await otbr_entry.runtime_data.set_active_dataset_tlvs(
```

Comment:
We shouldn't access another integrations runtime data if it's not exposed in an exported function.